### PR TITLE
[gym] support Commandline apps 💪💻 2/2

### PR DESCRIPTION
**Pull request by hjanuschka**
@ohwutup i don't know how i managed to smash the previous PR - i had a bad rebase, pushed force (thinking being on a different branch) and :boom: PR was closed, i didn't find a way to reopen it

BIG SORRY FOR THE TROUBLES - stupid me!

split of #7005 

test it with embeded `commandline` project.

```ruby
lane :powerlift do
	cmd_tool
	cmd_lib
	cmd_framework
	macapp
	touch_library
	touch_framework
	touch_app
        tvos
end

lane :tvos do
	 gym(scheme: "tvos",project: "gym/examples/commandline/commandline.xcodeproj", clean: false, output_directory: "out")	
end


lane :cmd_tool do
	 gym(scheme: "cmd-tool",project: "gym/examples/commandline/commandline.xcodeproj", clean: false, output_directory: "out")	
end

lane :cmd_lib do
	 gym(scheme: "cmd-library",project: "gym/examples/commandline/commandline.xcodeproj", clean: false, output_directory: "out")	
end


lane :cmd_framework do
	 gym( output_directory: "out", scheme: "cmd-framework",derived_data_path: "asdf", project: "gym/examples/commandline/commandline.xcodeproj", clean: false)	
end

lane :macapp do
	 gym(scheme: "macapp",project: "gym/examples/commandline/commandline.xcodeproj", clean: false)	
end

lane :touch_library do
	 gym( output_directory: "out", scheme: "touch-library",project: "gym/examples/commandline/commandline.xcodeproj", clean: false)	
end

lane :touch_framework do
	 gym( output_directory: "out", scheme: "touch-framework",project: "gym/examples/commandline/commandline.xcodeproj", clean: false)	
end

lane :touch_app do
	 gym(  output_directory: "out",scheme: "touch-app",project: "gym/examples/commandline/commandline.xcodeproj", clean: false)	
end
```

### CODE REVIEW

**Comment by milch** [comment]
Can you make the `.xcodeproj` a bit more barebones please? There are a couple user configuration files in there, along with (I think) unneeded asset catalogs and many duplicate files that could be reused (`main.m`, `ViewController.*`, etc.)

**Comment by hjanuschka** [comment]
@milch  - this is the most simple thing, i just created a new project, and added one target after another, without coding something - i am not sure what to remove.

or do you mean that the main.m should be linked in the target instead of a real file? (not sure if this is a good idea)

**Comment by hjanuschka** [comment]
@milch  ok i tried to remove a few things, like copyright, asset catalogs

**Comment by milch** [comment]
Great, thanks! 🚀

**Comment by KrauseFx** [comment]
Why did you try to sneak bin/🏋️  in? We already have an emoji alias for 💪

**Comment by milch** [comment]
It seems like Github can't handle reviews for files with emoji names (*outrage!*)

screen shot 2016-11-29 at 15 15 28

So anyway, let's also remove the 🏋️  alias. That should be a symlink, anyway. If you want you can submit a separate PR, but be sure to include 🏋️  and :weight_lifting_woman:!

**Comment by hjanuschka** [comment]
@milch  @KrauseFx  i thought of turning gym into a powerlifting tool.
however, i just removed it.

**Comment by milch** [comment]
@hjanuschka 

I just tried this PR again after merging the `fastlane_core` part, however, I seem to be running into the following issue:

### Captured Output

Command Used: `powerlift`
<details><summary>Output/Log</summary>

```

+------------------------+---------+---------------------------------+
|                            Used plugins                            |
+------------------------+---------+---------------------------------+
| Plugin                 | Version | Action                          |
+------------------------+---------+---------------------------------+
| fastlane-plugin-docker | 0.1.0   | docker_build                    |
|                        |         | docker_client                   |
|                        |         | docker_login                    |
|                        |         | docker_machine_client           |
|                        |         | docker_push                     |
|                        |         | ensure_docker_machine_available |
+------------------------+---------+---------------------------------+

[15:41:56]: -----------------------
[15:41:56]: --- Step: xcversion ---
[15:41:56]: -----------------------
[15:41:57]: Setting Xcode version to /Applications/Xcode-8.1-signed.app for all build steps
[15:41:57]: Driving the lane 'powerlift' 🚀
[15:41:57]: -------------------------------------
[15:41:57]: --- Step: Switch to cmd_tool lane ---
[15:41:57]: -------------------------------------
[15:41:57]: Cruising over to lane 'cmd_tool' 🚖
[15:41:57]: -----------------
[15:41:57]: --- Step: gym ---
[15:41:57]: -----------------
[15:41:57]: $ xcodebuild -list -project ../fastlane/gym/examples/commandline/commandline.xcodeproj
[15:41:58]: $ xcodebuild clean -showBuildSettings -scheme cmd-tool -project ../fastlane/gym/examples/commandline/commandline.xcodeproj

+----------------------+------------------------------------------------------------+
|                              Summary for gym 1.12.1                               |
+----------------------+------------------------------------------------------------+
| scheme               | cmd-tool                                                   |
| project              | ../fastlane/gym/examples/commandline/commandline.xcodeproj |
| clean                | false                                                      |
| output_directory     | out                                                        |
| destination          | generic/platform=macOS                                     |
| output_name          | App                                                        |
| silent               | false                                                      |
| use_legacy_build_api | false                                                      |
| buildlog_path        | ~/Library/Logs/gym                                         |
| xcode_path           | /Applications/Xcode-8.1-signed.app                         |
+----------------------+------------------------------------------------------------+

[15:41:58]: $ set -o pipefail && xcodebuild -scheme cmd-tool -project ../fastlane/gym/examples/commandline/commandline.xcodeproj -destination 'generic/platform=macOS' -archivePath gym/examples//Library/Developer/Xcode/Archives/2016-12-05/App\ 2016-12-05\ 15.41.58.xcarchive archive | tee gym/examples//Library/Logs/gym/App-cmd-tool.log | xcpretty
[15:41:59]: ▸ Building commandline/cmd-tool [Release]
[15:41:59]: ▸ Check Dependencies
[15:41:59]: ▸ Compiling main.m
[15:41:59]: ▸ Linking cmd-tool
[15:41:59]: ▸ Generating 'cmd-tool.dSYM'
[15:41:59]: ▸ Signing gym/examples//Library/Developer/Xcode/DerivedData/commandline-afrlbivrjjktbddsoeutodiobkjb/Build/Intermediates/ArchiveIntermediates/cmd-tool/InstallationBuildProductsLocation/usr/local/bin/cmd-tool
[15:41:59]: ▸ Touching cmd-tool.dSYM
[15:41:59]: ▸ Archive Succeeded
[15:41:59]: Successfully exported the binary files:
[15:41:59]: gym/examples//Library/Developer/Xcode/Archives/2016-12-05/App 2016-12-05 15.41.58.xcarchive/Products/usr/local/bin/cmd-tool
[15:41:59]: Cruising back to lane 'powerlift' 🚘
[15:41:59]: ------------------------------------
[15:41:59]: --- Step: Switch to cmd_lib lane ---
[15:41:59]: ------------------------------------
[15:41:59]: Cruising over to lane 'cmd_lib' 🚖
[15:41:59]: -----------------
[15:41:59]: --- Step: gym ---
[15:41:59]: -----------------
[15:41:59]: $ xcodebuild -list -project ../fastlane/gym/examples/commandline/commandline.xcodeproj
[15:42:00]: $ xcodebuild clean -showBuildSettings -scheme cmd-library -project ../fastlane/gym/examples/commandline/commandline.xcodeproj

+----------------------+------------------------------------------------------------+
|                              Summary for gym 1.12.1                               |
+----------------------+------------------------------------------------------------+
| scheme               | cmd-library                                                |
| project              | ../fastlane/gym/examples/commandline/commandline.xcodeproj |
| clean                | false                                                      |
| output_directory     | out                                                        |
| destination          | generic/platform=macOS                                     |
| output_name          | App                                                        |
| silent               | false                                                      |
| use_legacy_build_api | false                                                      |
| buildlog_path        | ~/Library/Logs/gym                                         |
| xcode_path           | /Applications/Xcode-8.1-signed.app                         |
+----------------------+------------------------------------------------------------+

[15:42:00]: $ set -o pipefail && xcodebuild -scheme cmd-library -project ../fastlane/gym/examples/commandline/commandline.xcodeproj -destination 'generic/platform=macOS' -archivePath gym/examples//Library/Developer/Xcode/Archives/2016-12-05/App\ 2016-12-05\ 15.42.00.xcarchive archive | tee gym/examples//Library/Logs/gym/App-cmd-library.log | xcpretty
[15:42:01]: ▸ Building commandline/cmd-library [Release]
[15:42:01]: ▸ Check Dependencies
[15:42:01]: ▸ Compiling library.m
[15:42:01]: ▸ Linking libcmd-library.dylib
[15:42:01]: ▸ Generating 'libcmd-library.dylib.dSYM'
[15:42:01]: ▸ Signing gym/examples//Library/Developer/Xcode/DerivedData/commandline-afrlbivrjjktbddsoeutodiobkjb/Build/Intermediates/ArchiveIntermediates/cmd-library/InstallationBuildProductsLocation/usr/local/lib/libcmd-library.dylib
[15:42:01]: ▸ Touching libcmd-library.dylib.dSYM
[15:42:01]: ▸ Archive Succeeded
[15:42:01]: Successfully exported the binary files:
[15:42:01]: gym/examples//Library/Developer/Xcode/Archives/2016-12-05/App 2016-12-05 15.42.00.xcarchive/Products/usr/local/lib/libcmd-library.dylib
+--------------------+-----------------------------------------------------------------------------+
|                                           Lane Context                                           |
+--------------------+-----------------------------------------------------------------------------+
| PLATFORM_NAME      |                                                                             |
| LANE_NAME          | powerlift                                                                   |
| DSYM_OUTPUT_PATH   | gym/examples//testproj/out                                                |
| XCODEBUILD_ARCHIVE | gym/examples//Library/Developer/Xcode/Archives/2016-12-05/App 2016-12-... |
+--------------------+-----------------------------------------------------------------------------+
[15:42:01]: File exists @ syserr_fail2_in - gym/examples//testproj/out/Release/libcmd-library.dylib

+------+-------------------------+-------------+
|               fastlane summary               |
+------+-------------------------+-------------+
| Step | Action                  | Time (in s) |
+------+-------------------------+-------------+
| 1    | xcversion               | 1           |
| 2    | Switch to cmd_tool lane | 0           |
| 3    | gym                     | 2           |
| 4    | Switch to cmd_lib lane  | 0           |
| 💥    | gym                     | 1           |
+------+-------------------------+-------------+

[15:42:01]: fastlane finished with errors

Looking for related GitHub issues on fastlane/fastlane...

Found no similar issues. To create a new issue, please visit:
https://github.com/fastlane/fastlane/issues/new
Run `fastlane env` to append the fastlane environment to your issue


```

</details>

<details><summary>✅ fastlane environment ✅</summary>

### Stack

| Key                 | Value                                                    |
| ------------------- | -------------------------------------------------------- |
| OS                  | 10.11.6                                                  |
| Ruby                | 2.3.1                                                    |
| Bundler?            | true                                                     |
| Git                 | git version 2.10.2                                       |
| Installation Source | ~/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/bin/fastlane |
| Host                | Mac OS X 10.11.6 (15G1108)                               |
| Ruby Lib Dir        | ~/.rbenv/versions/2.3.1/lib                              |
| OpenSSL Version     | OpenSSL 1.0.2h  3 May 2016                               |
| Is contained        | false                                                    |
| Xcode Path          | /Applications/Xcode-8.1-signed.app/Contents/Developer/   |
| Xcode Version       | 8.1                                                      |


### System Locale

| Variable | Value       |   |
| -------- | ----------- | - |
| LANG     | de_AT.UTF-8 | ✅ |
| LC_ALL   | de_AT.UTF-8 | ✅ |
| LANGUAGE |             |   |


### fastlane files:

<details><summary>`./fastlane/Fastfile`</summary>

```ruby
xcversion(version: "8.1")

lane :powerlift do
	cmd_tool
	cmd_lib
	cmd_framework
	macapp
	touch_library
	touch_framework
	touch_app
        tvos
end

lane :tvos do
	 gym(scheme: "tvos",project: "../fastlane/gym/examples/commandline/commandline.xcodeproj", clean: false, output_directory: "out")	
end


lane :cmd_tool do
	 gym(scheme: "cmd-tool",project: "../fastlane/gym/examples/commandline/commandline.xcodeproj", clean: false, output_directory: "out")	
end

lane :cmd_lib do
	 gym(scheme: "cmd-library",project: "../fastlane/gym/examples/commandline/commandline.xcodeproj", clean: false, output_directory: "out")	
end


lane :cmd_framework do
	 gym( output_directory: "out", scheme: "cmd-framework",derived_data_path: "asdf", project: "../fastlane/gym/examples/commandline/commandline.xcodeproj", clean: false)	
end

lane :macapp do
	 gym(scheme: "macapp",project: "../fastlane/gym/examples/commandline/commandline.xcodeproj", clean: false)	
end

lane :touch_library do
	 gym( output_directory: "out", scheme: "touch-library",project: "../fastlane/gym/examples/commandline/commandline.xcodeproj", clean: false)	
end

lane :touch_framework do
	 gym( output_directory: "out", scheme: "touch-framework",project: "../fastlane/gym/examples/commandline/commandline.xcodeproj", clean: false)	
end

lane :touch_app do
	 gym(  output_directory: "out",scheme: "touch-app",project: "../fastlane/gym/examples/commandline/commandline.xcodeproj", clean: false)	
end
```
</details>

**No Appfile found**


### fastlane gems

| Gem                 | Version | Update-Status |
| ------------------- | ------- | ------------- |
| credentials_manager | 0.16.2  | ✅ Up-To-Date  |
| fastlane_core       | 0.58.0  | ✅ Up-To-Date  |
| spaceship           | 0.38.5  | ✅ Up-To-Date  |
| cert                | 1.4.4   | ✅ Up-To-Date  |
| deliver             | 1.16.1  | ✅ Up-To-Date  |
| frameit             | 3.0.0   | ✅ Up-To-Date  |
| gym                 | 1.12.1  | ✅ Up-To-Date  |
| sigh                | 1.12.1  | ✅ Up-To-Date  |
| match               | 0.11.1  | ✅ Up-To-Date  |
| pem                 | 1.4.1   | ✅ Up-To-Date  |
| pilot               | 1.12.1  | ✅ Up-To-Date  |
| produce             | 1.3.2   | ✅ Up-To-Date  |
| scan                | 0.14.2  | ✅ Up-To-Date  |
| screengrab          | 0.5.6   | ✅ Up-To-Date  |
| snapshot            | 1.16.4  | ✅ Up-To-Date  |
| supply              | 0.8.0   | ✅ Up-To-Date  |
| fastlane            | 1.111.0 | ✅ Up-To-Date  |


### Loaded fastlane plugins:

| Plugin                 | Version | Update-Status |
| ---------------------- | ------- | ------------- |
| fastlane-plugin-docker | 0.1.0   | ✅ Up-To-Date  |


<details><summary><b>Loaded gems</b></summary>

| Gem                       | Version      |
| ------------------------- | ------------ |
| did_you_mean              | 1.0.2        |
| bundler                   | 1.13.6       |
| io-console                | 0.4.6        |
| CFPropertyList            | 2.3.4        |
| i18n                      | 0.7.0        |
| json                      | 1.8.3        |
| minitest                  | 5.10.1       |
| thread_safe               | 0.3.5        |
| tzinfo                    | 1.2.2        |
| activesupport             | 4.2.7.1      |
| public_suffix             | 2.0.4        |
| addressable               | 2.5.0        |
| babosa                    | 1.0.2        |
| colored                   | 1.2          |
| highline                  | 1.7.8        |
| commander                 | 4.4.1        |
| security                  | 0.1.3        |
| credentials_manager       | 0.16.2       |
| excon                     | 0.54.0       |
| gh_inspector              | 1.0.2        |
| multi_json                | 1.12.1       |
| plist                     | 3.2.0        |
| rubyzip                   | 1.1.7        |
| terminal-table            | 1.4.5        |
| fastlane_core             | 0.58.0       |
| multipart-post            | 2.0.0        |
| faraday                   | 0.10.0       |
| unf_ext                   | 0.0.7.2      |
| unf                       | 0.1.4        |
| domain_name               | 0.5.20161129 |
| http-cookie               | 1.0.3        |
| faraday-cookie_jar        | 0.0.6        |
| faraday_middleware        | 0.10.1       |
| fastimage                 | 1.6.8        |
| multi_xml                 | 0.5.5        |
| claide                    | 1.0.1        |
| coderay                   | 1.1.1        |
| dotenv                    | 2.1.1        |
| mini_magick               | 4.5.1        |
| rouge                     | 1.11.1       |
| xcpretty                  | 0.2.4        |
| net-ssh                   | 3.2.0        |
| net-sftp                  | 2.1.2        |
| krausefx-shenzhen         | 0.14.11      |
| slack-notifier            | 1.5.1        |
| xcpretty-travis-formatter | 0.0.4        |
| jwt                       | 1.5.6        |
| little-plugger            | 1.1.4        |
| logging                   | 2.1.0        |
| memoist                   | 0.15.0       |
| os                        | 0.9.6        |
| signet                    | 0.7.3        |
| googleauth                | 0.5.1        |
| httpclient                | 2.8.2.4      |
| hurley                    | 0.2          |
| mime-types-data           | 3.2016.0521  |
| mime-types                | 3.1          |
| uber                      | 0.0.15       |
| representable             | 2.3.0        |
| retriable                 | 2.1.0        |
| google-api-client         | 0.9.20       |
| terminal-notifier         | 1.7.1        |
| word_wrap                 | 1.0.0        |
| xcode-install             | 2.0.9        |
| nanaimo                   | 0.2.3        |
| xcodeproj                 | 1.4.1        |
| fastlane-plugin-docker    | 0.1.0        |
| method_source             | 0.8.2        |
| slop                      | 3.6.0        |
| pry                       | 0.10.4       |
| watchbuild                | 0.1.4        |
</details>


*generated on:* **2016-12-05**
</details>

**Comment by hjanuschka** [comment]
@milch  sorry for that!
i actually figured out what the problem was, and refactored a few things, i had to rebase the PR (sorry for that) - the changes since last push are here: https://github.com/hjanuschka/fastlane/commit/c0dfb30229f88ffdc7071e091747beac91fbfcf4

the main problem was with symlinks, and frameworks/libs do not build final products (resulting in broken symlinks) - command_builder now does `build` instead of `archive`  on libs/frameworks.

and the whole copy/extract part has been redone.

File: gym/lib/gym/runner.rb
Comment (milch) [inline gym/lib/gym/runner.rb]: Hmm, do we have to do a recursive delete here? Just wanting to make sure that we don't delete too many or the wrong things by mistake, e.g. if somehow `~` ends up in there by some strange miracle.

File: gym/lib/gym/runner.rb
Comment (milch) [inline gym/lib/gym/runner.rb]: Return this explicitly if you need it, otherwise remove please (and get rid of the `framework_path` temporary)

**Review by milch — APPROVED** [review]
Alright, thanks for your changes! Just a couple notes here left, then we should be good to merge 🚀

File: gym/lib/gym/runner.rb
Comment (hjanuschka) [inline gym/lib/gym/runner.rb]: tbh. i was suprised by the bad beahviour on `cp_r` when there are symlinks in and subfolder. if the subfolder already exists they where not overwriteable.  i even tried to shell out, but osx itself has problems with recursive copy and symlink dereferencing. i found a few ruby hacks, that use `find` - but i think thats way too ugly.

the current solution uses:
`existing_file = File.join(File.expand_path(Gym.config[:output_directory]), File.basename(f))`

whereas there is a small chance for an edge-case where the user sets output_directory to `/` (not sure if this even works on osx - without some kernel params) - BUT `f` should never be empty, as `f`  is a result of the each loop. i also think - that of other FS restricitions there should be no chance that `f`  could be `*,~` or some other system relevant value.

let me know if you have another opinion - to further improve it!

File: gym/lib/gym/runner.rb
Comment (hjanuschka) [inline gym/lib/gym/runner.rb]: removed.

File: gym/lib/gym/runner.rb
Comment (hjanuschka) [inline gym/lib/gym/runner.rb]: now `user_error!` 👍  thx!

**Review by hjanuschka — COMMENTED** [review]
@milch thx - feedback addressed
really looking forward to this, i think this is a nice progress for gym!

File: gym/lib/gym/runner.rb
Comment (milch) [inline gym/lib/gym/runner.rb]: Hmm, also I'm guessing that since `File.basename(f)` is not expanded, it would only try to delete a file called `~` or `*` which obviously wouldn't exist. I'm gonna leave this then.

File: gym/lib/gym/generators/build_command_generator.rb
Comment (KrauseFx) [inline gym/lib/gym/generators/build_command_generator.rb]: It seems like we always want either `archive` or `build` here. If so then the if and unless blocks are wrong 😜

Comment (KrauseFx) [inline gym/lib/gym/generators/build_command_generator.rb]: yeah, seems like this was a very basic computer science error 😜

Comment (hjanuschka) [inline gym/lib/gym/generators/build_command_generator.rb]: @KrauseFx  😨 shame on me! sorry - i don't know how it even worked when testing the sample projects :(

**Pull request merged**
